test/requestContextPlugin.e2e.spec.js

```diff
@@ -0,0 +1,106 @@
+const fastify = require('fastify')
+const request = require('superagent')
+const { fastifyRequestContextPlugin } = require('../lib/requestContextPlugin')
+const { TestService } = require('./internal/testService')
+
+function initAppPostWithPrevalidation(endpoint) {
+  return new Promise((resolve) => {
+    const app = fastify({ logger: true })
+    app.register(fastifyRequestContextPlugin, { hook: 'preValidation' })
+
+    const preValidationFn = (req, reply, done) => {
+      const requestId = Number.parseInt(req.body.requestId)
+      req.requestContext.set('testKey', `testValue${requestId}`)
+      done()
+    }
+
+    app.route({
+      url: '/',
+      method: ['GET', 'POST'],
+      preValidation: preValidationFn,
+      handler: endpoint,
+    })
+
+    app.listen(8085, '0.0.0.0', (err, address) => {
+      if (err) {
+        console.warn(err)
+        process.exit(1)
+      }
+      console.info(`Server listening at ${address}`)
+      resolve(app)
+    })
+  })
+}
+
+describe('requestContextPlugin E2E', () => {
+  let app
+  afterEach(() => {
+    return app.close()
+  })
+
+  it('correctly preserves values set in prevalidation phase within single POST request', () => {
+    expect.assertions(2)
+
+    let testService
+    let responseCounter = 0
+    return new Promise((resolveResponsePromise) => {
+      const promiseRequest2 = new Promise((resolveRequest2Promise) => {
+        const promiseRequest1 = new Promise((resolveRequest1Promise) => {
```

Comment:
please use async functions instead.

Reply:
Are you sure that would work as expected? In this case we want to able to have flexibility to control when each promise resolves specifically, and I don't think that is possible with async functions.

Reply:
oh ok.